pkg/scanners/trivy/trivy.go

```diff
@@ -0,0 +1,435 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	machinerytypes "k8s.io/apimachinery/pkg/types"
+
+	eraserv1alpha1 "github.com/Azure/eraser/api/v1alpha1"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/Azure/eraser/pkg/logger"
+	"github.com/aquasecurity/fanal/analyzer/config"
+	"github.com/aquasecurity/fanal/applier"
+	"github.com/aquasecurity/fanal/artifact"
+	artifactImage "github.com/aquasecurity/fanal/artifact/image"
+	"github.com/aquasecurity/fanal/cache"
+	fanalImage "github.com/aquasecurity/fanal/image"
+	fanalTypes "github.com/aquasecurity/fanal/types"
+	"github.com/aquasecurity/trivy-db/pkg/db"
+	dlDb "github.com/aquasecurity/trivy/pkg/db"
+	"github.com/aquasecurity/trivy/pkg/detector/ospkg"
+	pkgResult "github.com/aquasecurity/trivy/pkg/result"
+	"github.com/aquasecurity/trivy/pkg/scanner"
+	"github.com/aquasecurity/trivy/pkg/scanner/local"
+	trivyTypes "github.com/aquasecurity/trivy/pkg/types"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	generalErr = 1
+
+	severityCritical = "CRITICAL"
+	severityHigh     = "HIGH"
+	severityMedium   = "MEDIUM"
+	severityLow      = "LOW"
+	severityUnknown  = "UNKNOWN"
+
+	vulnTypeOs      = "os"
+	vulnTypeLibrary = "library"
+
+	securityCheckVuln   = "vuln"
+	securityCheckConfig = "config"
+	securityCheckSecret = "secret"
+
+	apiPath         = "apis/eraser.sh/v1alpha1"
+	resourceName    = "imagecollectors"
+	subResourceName = "status"
+	trivyVersion    = "v0.24.4"
```

Comment:
i am not familiar with this, does this need to match the go.mod version of trivy?

Reply:
No, not necessarily, but it probably should. Could we set this using LDFlags based on the version in `go.mod`?

Reply:
Looks like trivy gets this info from appVersion, can we use that or is that part of cli only? https://github.com/aquasecurity/trivy/blob/1b66b77f691f2249a63c284056afcad704c0e15f/pkg/commands/option/global.go#L35

Reply:
@sozercan I think it's CLI only. I am setting this using ldflags in the makefile:
```
TRIVY_VERSION ?= $(shell go list -f '{{ .Version }}' -m github.com/aquasecurity/trivy)
...
TRIVY_SCANNER_LDFLAGS ?= $(LDFLAGS)-X 'main.trivyVersion=$(TRIVY_VERSION)'
```

Reply:
@sozercan This is only used for error output. Trivy's process for fetching db looks like this:
- Look at the db file on disk, if the db file doesn't exist yet, then this is the first run and it will automatically fetch the latest version of the db, using an OCI client, from `ghcr`. It uses major version tags to fetch the latest version
- the `trivy` library depends on the `trivy-db` package, and that package has a constant `db.Version` set to the major version of the db
- Internally, if the `db.Version` is compared against the `SchemaVersion` found in the db on the filesystem. If `db.Version` is less than `SchemaVersion`, it means we are using an older client with a newer db, and `trivyVersion` will be propagated into the error output.
- If, `db.Version` is >= `SchemaVersion`, then `db.Version` (the version determined by the `trivy-db` version, which is determined by the `trivy` version) will be used as the major version tag for the OCI fetch.
- Finally, timestamps are compared to determine whether or not the db needs to be updated or not.